crates/autopilot/src/boundary/events/cow_amm_product_factory.rs

```diff
@@ -0,0 +1,67 @@
+use {
+    contracts::cow_amm_constant_product_factory::Event,
+    ethrpc::current_block::RangeInclusive,
+    shared::{event_handling::EventStoring, impl_event_retrieving},
+    std::collections::BTreeMap,
+};
+
+impl_event_retrieving! {
+    pub CowAmmConstantProductFactoryContract for contracts::cow_amm_constant_product_factory
+}
+
+#[derive(Default)]
+/// CoW AMM indexer which stores events in-memory.
+pub struct Indexer {
+    /// Registry with the format: (block number, event)
+    registry: BTreeMap<u64, contracts::cow_amm_constant_product_factory::event_data::Deployed>,
+}
+
+#[async_trait::async_trait]
+impl EventStoring<contracts::cow_amm_constant_product_factory::Event> for Indexer {
+    async fn replace_events(
+        &mut self,
+        events: Vec<ethcontract::Event<contracts::cow_amm_constant_product_factory::Event>>,
+        range: RangeInclusive<u64>,
+    ) -> anyhow::Result<()> {
+        // Remove events in the specified range
+        let range = *range.start()..=*range.end();
+        let keys_to_remove: Vec<u64> = self
+            .registry
+            .range(range)
+            .map(|(block, _)| *block)
+            .collect();
+
+        for key in keys_to_remove {
+            self.registry.remove(&key);
+        }
+
+        self.append_events(events).await
+    }
+
+    async fn append_events(
+        &mut self,
+        events: Vec<ethcontract::Event<contracts::cow_amm_constant_product_factory::Event>>,
+    ) -> anyhow::Result<()> {
+        events.into_iter().try_for_each(|event| {
+            let block_number = event
+                .meta
+                .ok_or_else(|| anyhow::anyhow!("Event missing meta"))?
+                .block_number;
+            match event.data {
+                Event::Deployed(event) => {
+                    self.registry.insert(block_number, event);
+                }
+                _ => {}
+            }
+            Ok::<_, anyhow::Error>(())
+        })?;
+        Ok(())
+    }
+
+    async fn last_event_block(&self) -> anyhow::Result<u64> {
+        self.registry
+            .last_key_value()
+            .map(|(block, _)| *block)
+            .ok_or_else(|| anyhow::anyhow!("No events found"))
```

Comment:
This function gets used to let the indexer know where it should start indexing to get the missing events. This should either return 0 or the block where the factory got created (not sure if the indexer already starts at the creation block at the earliest).
I think you probably have to add the creation block for the factory contract in the `contracts/build.rs` file where we can specify some metadata on all the contract bindinds.

Reply:
Thanks a lot! What do you exactly mean by
> I think you probably have to add the creation block for the factory contract in the contracts/build.rs file where we can specify some metadata on all the contract bindinds.